marlowe-playground-client/src/Component/MetadataTab/View.purs

```diff
@@ -209,38 +227,57 @@ metadataList metadataAction metadataMap hintSet metadataRenderer typeNameTitle t
       (Map.fromFoldable (map (\x -> x /\ Nothing) ((toUnfoldable hintSet) :: List String)))
 
 sortableMetadataList ::
-  forall a b c p.
-  (b -> a) ->
+  forall a c p r.
+  { moveDown :: String -> a
+  , moveUp :: String -> a
+  , set :: String -> String -> a
+  | r
+  } ->
   OMap String c ->
   OSet String ->
-  (String -> c -> Boolean -> (b -> a) -> String -> String -> Array (HTML p a)) ->
-  String -> String -> (String -> b) -> Array (HTML p a)
-sortableMetadataList metadataAction metadataMap hintSet metadataRenderer typeNameTitle typeNameSmall setEmptyMetadata =
+  (String -> c -> Boolean -> String -> String -> Array (HTML p a)) ->
+  String -> String -> Array (HTML p a)
+sortableMetadataList actions metadataMap hintSet metadataRenderer typeNameTitle typeNameSmall = do
+  let
+    items = OMap.toUnfoldable combinedMap
+
+    itemsCount = Array.length items
   if OMap.isEmpty combinedMap then
     []
   else
-    [ div [ class_ $ ClassName "metadata-group-title" ]
-        [ h6_ [ em_ [ text $ typeNameTitle <> " descriptions" ] ] ]
-    ]
-      <> ( concatMap
-            ( \(key /\ val) ->
-                ( case val of
-                    Just (info /\ needed) -> metadataRenderer key info needed metadataAction typeNameTitle typeNameSmall
-                    Nothing ->
-                      [ div [ classes [ ClassName "metadata-error", ClassName "metadata-prop-not-defined" ] ]
-                          [ text $ typeNameTitle <> " " <> show key <> " meta-data not defined" ]
-                      , div [ class_ $ ClassName "metadata-prop-create" ]
-                          [ button
-                              [ classes [ minusBtn, ClassName "align-top", btn ]
-                              , onClick $ const $ metadataAction (setEmptyMetadata key)
-                              ]
-                              [ text "+" ]
-                          ]
-                      ]
-                )
-            )
-            $ OMap.toUnfoldable combinedMap
-        )
+    AB.build
```

Comment:
`ArrayBuilder` - nice, I'm guesing this is like a difference list that defers all updates until the end then "fuses" them together mutably in JS? That's really handy.

Reply:
Yeah. It closely follows `Record.Builder` internals.